src/System.Private.CoreLib/shared/System/Double.cs

```diff
@@ -226,7 +226,7 @@ public bool Equals(double obj)
         [MethodImpl(MethodImplOptions.AggressiveInlining)] // 64-bit constants make the IL unusually large that makes the inliner to reject the method
         public override int GetHashCode()
         {
-            var bits = Unsafe.As<double, long>(ref m_value);
+            var bits = BitConverter.DoubleToInt64Bits(m_value);
```

Comment:
This will make the GetHashCode method significantly slower, I think.

Reply:
Really? The BitConverter.Int64BitsToDouble() method is simply: 


https://github.com/dotnet/coreclr/blob/dd74fca452ccfc8666d4f88a19b79ffa3aa53d3d/src/mscorlib/shared/System/BitConverter.cs#L450


Which seems more straightforward than Unsafe.As(). Any way to review the generated assembly?

Reply:
> Which seems more straightforward than Unsafe.As().

It's not as `Int64BitsToDouble` will result in the `double` value being loaded as `double`, stored to a temporary and then loaded again as `long`. The `Unsafe.As` variant simply converts `this` (which is basically a `ref` in the case of structs) from `ref double` to `ref long`, that is a no-op.

Reply:
@mikedn Can you suggest how to get Unsafe.AsRef to work? I couldn't figure it out.

Reply:
Side questions: Can the compiler optimize away the cast by promoting the temporary into a register? And is there a X86/X64 instruction that can bit-copy from a floating point XMM register to a general purpose register?

Reply:
Add it to these places:

https://github.com/dotnet/coreclr/blob/master/src/vm/jitinterface.cpp#L7050
https://github.com/dotnet/coreclr/blob/master/src/vm/mscorlib.h#L778
https://github.com/dotnet/coreclr/blob/master/src/System.Private.CoreLib/shared/Internal/Runtime/CompilerServices/Unsafe.cs#L355

And then use it like this: `var bits = Unsafe.As<double, long>(ref Unsafe.AsRef(in m_value));`

> Can the compiler optimize away the cast by promoting the temporary into a register?

It can, but it is not exactly trivial optimization to build.

Reply:
@jkotas I made the change to Unsafe.cs; however, I'm not exactly sure what to do with jitinterface.cpp and mscorlib.h.

Reply:
Here you go

Reply:
We'll need to check if this approach doesn't impact code gen as well. Sometimes the JIT has problems with multiple calls - even if they are inlined it may still spill trees and introduce new local variables.

I'll try to take a look later today.

Reply:
Works fine, the JIT generates identical IR for both old and new implementations.

And it turns out that the original `As` call does result in a temporary variable being introduced. Fortunately it has no impact on the generated code.

Reply:
@mikedn Thanks for taking a look!